### PR TITLE
186 dice loss support multi-labels task

### LINKED ISSUE #186: dice loss support multi-labels task
**Is your feature request related to a problem? Please describe.**
The current implementation of DiceLoss can only support Binary and Multi-classes tasks.
Need to also support Multi-labels task.

### PR DESCRIPTION
**Pull request by Nic-Ma**
Fixes #186 .

### Description
This PR refactor the DiceLoss modules to support both multi-classes and multi-labels tasks.
And also update the "segmentation_prediction" tool.

### Status
**Ready**

### Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] Breaking change (fix or new feature that would cause existing functionality to change)
- [x] New tests added to cover the changes
- [x] Docstrings/Documentation updated

### CODE REVIEW

File: monai/losses/dice.py [hunk 1]
```diff
@@ -23,63 +23,82 @@
 @alias("dice", "Dice")
 class DiceLoss(_Loss):
     """
-    Multiclass dice loss. Input logits `pred` (BNHW[D] where N is number of classes) is compared with ground truth
```
Comment (wyli) [inline monai/losses/dice.py]: would be great to revise the docstring instead of removing it

File: monai/losses/dice.py [hunk 2]
```diff
@@ -23,63 +23,82 @@
...
+        """
+        Args:
+            pred (tensor): the shape should be BNH[WD].
+            ground (tensor): the shape should be BNH[WD].
+            smooth (float): a small constant to avoid nan.
+        """
+        n_pred_ch = pred.shape[1]
+        if self.do_sigmoid is True:
+            pred = torch.sigmoid(pred)
+        if self.do_softmax is True:
```
Comment (wyli) [inline monai/losses/dice.py]: when both do_softmax=True and do_sigmoid =True, it should raise a ValueError

File: monai/losses/dice.py [hunk 3]
```diff
@@ -121,44 +149,42 @@ def forward(self, pred, ground, smooth=1e-5):
...
-            psum = psum.sigmoid()  # use sigmoid activation
-        if pred.shape[1] == 1:
-            if self.do_softmax:
-                raise ValueError('do_softmax is not compatible with single channel prediction.')
-            if not self.include_background:
+        n_pred_ch = pred.shape[1]
+        if self.do_sigmoid is True:
+            pred = torch.sigmoid(pred)
+        if self.do_softmax is True:
+            pred = torch.softmax(pred, 1)
```
Comment (wyli) [inline monai/losses/dice.py]: as my previous commnet, do_sigmoid=True self.do_softmax=True not compatible. `if self.do_sigmid is True:` could be `if self.do_sigmoid:` (the same for the other places)

File: monai/losses/dice.py [hunk 4]
```diff
@@ -23,63 +23,82 @@
...
-            tsum = ground
-        else:  # multiclass dice loss
-            if self.do_softmax:
-                psum = torch.softmax(pred, 1)
-            tsum = one_hot(ground, pred.shape[1])  # B1HW(D) -> BNHW(D)
-            # exclude background category so that it doesn't overwhelm the other segmentations if they are small
-            if not self.include_background:
-                tsum = tsum[:, 1:]
-                psum = psum[:, 1:]
-        assert tsum.shape == psum.shape, ("Ground truth one-hot has differing shape (%r) from source (%r)" %
```
Comment (wyli) [inline monai/losses/dice.py]: i think this shape check is informative shouldn't be removed

File: monai/losses/dice.py [hunk 5]
```diff
@@ -23,63 +23,82 @@
...
+            ground (tensor): the shape should be BNH[WD].
+            smooth (float): a small constant to avoid nan.
+        """
+        n_pred_ch = pred.shape[1]
+        if self.do_sigmoid is True:
+            pred = torch.sigmoid(pred)
+        if self.do_softmax is True:
+            pred = torch.softmax(pred, 1)
+        if self.to_onehot_y is True:
+            if n_pred_ch == 1:
```
Comment (wyli) [inline monai/losses/dice.py]: `softmax` should be ignored if it's one class prediction. I think we'll have better readability if all the options for one-class prediction is handled within a block of `if n_pred_ch==1:` in one place

File: monai/networks/utils.py [hunk 6]
```diff
@@ -46,12 +46,19 @@ def slice_channels(tensor, *slicevals):
...
+
+    Args:
+        logits (Tensor): raw data of model output.
+        mutually_exclusive (Bool): if True, `logits` will be converted into a binary matrix using
+            a combination of argmax, which is suitable for multi-classes task. Defaults to False.
+        threshold (float): thresholding the prediction values if multi-labels task.
     """
-    if logits.shape[1] == 1:
-        return (logits >= 0).int()  # for binary segmentation threshold on channel 0
+    if mutually_exclusive is False:
```
Comment (wyli) [inline monai/networks/utils.py]: `if not mutually_exclusive`

File: monai/networks/utils.py [hunk 7]
```diff
@@ -46,12 +46,19 @@ def slice_channels(tensor, *slicevals):
...
+    if multi-labels task, computing the `argmax` along the channel axis if multi-classes task,
+    logits has shape `BCHW[D]`.
+
+    Args:
+        logits (Tensor): raw data of model output.
+        mutually_exclusive (Bool): if True, `logits` will be converted into a binary matrix using
+            a combination of argmax, which is suitable for multi-classes task. Defaults to False.
+        threshold (float): thresholding the prediction values if multi-labels task.
     """
-    if logits.shape[1] == 1:
```
Comment (wyli) [inline monai/networks/utils.py]: warns if channel==1 and mutuall_exclusive

File: monai/networks/utils.py [hunk 8]
```diff
@@ -46,12 +46,19 @@ def slice_channels(tensor, *slicevals):
...
-        return (logits >= 0).int()  # for binary segmentation threshold on channel 0
+    if mutually_exclusive is False:
+        return (logits >= threshold).int()
     else:
-        return logits.argmax(1).unsqueeze(1)  # take the index of the max value along dimension 1
+        return logits.argmax(1).unsqueeze(1)
```
Comment (wyli) [inline monai/networks/utils.py]: could use `argmax(1, keepdim=True)`

File: monai/losses/dice.py (hunk 1 above)
Comment (Nic-Ma) [reply monai/losses/dice.py]: updated in the latest patch.
Thanks.

File: monai/losses/dice.py (hunk 2 above)
Comment (Nic-Ma) [reply monai/losses/dice.py]: checked in __init__().
Thanks.

File: monai/losses/dice.py (hunk 3 above)
Comment (Nic-Ma) [reply monai/losses/dice.py]: updated in the latest patch.
Thanks.

File: monai/losses/dice.py (hunk 4 above)
Comment (Nic-Ma) [reply monai/losses/dice.py]: Added in the latest patch.
Thanks.

File: monai/losses/dice.py (hunk 5 above)
Comment (Nic-Ma) [reply monai/losses/dice.py]: updated in the latest patch.
Thanks.

File: monai/networks/utils.py (hunk 6 above)
Comment (Nic-Ma) [reply monai/networks/utils.py]: updated in the latest patch.
Thanks.

File: monai/networks/utils.py (hunk 8 above)
Comment (Nic-Ma) [reply monai/networks/utils.py]: updated in the latest patch.
Thanks.

File: monai/networks/utils.py (hunk 7 above)
Comment (Nic-Ma) [reply monai/networks/utils.py]: updated in the latest patch.
Thanks.

**Comment by Nic-Ma** [comment]
Hi @wyli ,

Thanks for your review.
I update the code according to your suggestions.
Could you please help review it again?
Thanks.
